### PR TITLE
Improve invocation_defects_tests with assertions and comments

### PR DESCRIPTION
**Pull request by lidizheng**
Fixes https://github.com/grpc/grpc/issues/21998

### CODE REVIEW

File: src/python/grpcio_tests/tests/unit/_invocation_defects_test.py [hunk 1]
```diff
@@ -199,6 +177,7 @@ def _defective_handler_multi_callable(channel):
 
 
 class InvocationDefectsTest(unittest.TestCase):
+    """Tests the handling of exception-raising user code."""
```
Comment (gnossen) [inline src/python/grpcio_tests/tests/unit/_invocation_defects_test.py]: Can we explicitly add "on the client side" to this comment?

Comment (lidizheng) [reply src/python/grpcio_tests/tests/unit/_invocation_defects_test.py]: Done.

**Comment by mehrdada** [comment]
Can we change the `request_iterator` to a generator that raises an exception explicitly as opposed to `next(x)` failing because user passed an iterable instead of an iterator (arguably should not be an error IMO).

**Comment by lidizheng** [comment]
@mehrdada That test case is covered by `testIteratorStreamRequestStreamResponse`. The failing is not generated by `next(x)` (response iterator), but inside the request iterator consumer. (I agree that we should not be too restrict.)

**Comment by mehrdada** [comment]
@lidizheng In that case, can we replace request_iterator with completely garbage input that does not support `__iter__` at all (i.e. not an iterable nor iterator) and call `iter(request_iterator)` in channel before `next`?

**Comment by lidizheng** [comment]
@mehrdada Done.

**Pull request merged**